Slot migration improvement

PingXie opened the pull request:
This was moved to #445 due to confilct.

## Overview
This PR significantly enhances the reliability and automation of the Valkey cluster re-sharding process, specifically during slot migrations in the face of primary failures. These updates address critical failure issues that previously required extensive manual intervention and could lead to data loss or inconsistent cluster states.

## Enhancements
### Automatic Failover Support in Empty Shards 
The cluster now supports automatic failover in shards that do not own any slots, which is common during scaling operations. This improvement ensures high availability and resilience from the outset of shard expansion.
### Replication of Slot Migration States
All `CLUSTER SETSLOT` commands are now initially executed on replica nodes before the primary. This ensures that the slot migration state is consistent within the shard, preventing state loss in the event of primary failure. A new timeout parameter has been introduced, allowing users to specify the duration in milliseconds to wait for replication to complete, with a default set at 2 seconds.
```
CLUSTER SETSLOT slot { IMPORTING node-id | MIGRATING node-id | NODE node-id | STABLE } [ TIMEOUT timeout ]
```
### Recovery of Logical Migration Links
The update automatically repairs the logical links between source and target nodes during failovers. This ensures that requests are correctly redirected to the new primary in the target shard after a primary failure, maintaining cluster integrity.
### Enhanced Support for New Replicas
New replicas added to shards involved in slot migrations will now automatically inherit the slot's migration state as part of their initialization. This ensures that new replicas are immediately consistent with the rest of the shard.
### Improved Logging for Slot Migrations
Additional logging has been implemented to provide operators with clearer insights into the slot migration processes and automatic recovery actions, aiding in monitoring and troubleshooting.


## Additional Changes

### cluster-allow-replica-migration

When `cluster-allow-replica-migration` is disabled, primary nodes that lose their last slot to another shard will no longer automatically become replicas of the receiving shard. Instead, they will remain in their own shards, which will now be empty, having no slots assigned to them.

Fix #21

Commit:
Slot migration improvement

Signed-off-by: Ping Xie <pingxie@google.com>

Commit:
Add missing changes

Signed-off-by: Ping Xie <pingxie@google.com>

Commit:
Improve test reliability

Signed-off-by: Ping Xie <pingxie@google.com>

PingXie commented:
@valkey-io/core-team ready for your review

zuiderkwast commented:
@PingXie I updated the PR description. Please check if it's correct and edit again if it's not.

zuiderkwast commented on src/valkey-cli.c:
```diff
@@ -4762,6 +4762,39 @@ static clusterManagerNode *clusterManagerGetSlotOwner(clusterManagerNode *n,
+                                                 clusterManagerNode *node2,
+                                                 int slot) {
+    /* A new CLUSTER SETSLOT variant that finalizes slot ownership on replicas
+     * only (CLUSTER SETSLOT s NODE n REPLICAONLY) is introduced in Redis 8.0+
```
```suggestion
     * only (CLUSTER SETSLOT s NODE n REPLICAONLY) is introduced in Valkey 8.0
```

zuiderkwast commented on src/valkey-cli.c:
```diff
@@ -4762,6 +4762,39 @@ static clusterManagerNode *clusterManagerGetSlotOwner(clusterManagerNode *n,
+     * ownership finalization on HA clusters. We make a best-effort attempt below
+     * to utilize this enhanced reliability. Regardless of the result, we continue
+     * with finalizing slot ownership on the primary nodes. Note that this command
+     * is not essential. Redis 8.0+ will attempt to recover from failed slot
```
```suggestion
     * is not essential. Valkey 8.0 will attempt to recover from failed slot
```

zuiderkwast commented on valkey.conf:
```diff
@@ -1672,8 +1672,10 @@ aof-timestamp-enabled no
 # Turning off this option allows to use less automatic cluster configuration.
-# It both disables migration to orphaned masters and migration from masters
-# that became empty.
+# It disables migration of replicas to orphaned masters. Masters that become
+# empty due to losing their last slots to another master will not automatically
+# replicate from the master that took over their last slots. Instead, they will
+# remain as empty masters without any slots.
```
For the release notes I suppose. Mention in PR description.

PingXie replied:
yeah, the "yes" behavior is the same but the "no" behavior is now deterministic that the empty primary won't automatically join the shard that takes away its last slot.

zuiderkwast reviewed (commented):
Partial review. (I've already reviewed it earlier and approved it, but it's long ago.)

PingXie committed:
Update src/valkey-cli.c

Co-authored-by: Viktor Söderqvist <viktor.soderqvist@est.tech>
Signed-off-by: Ping Xie <pingxie@google.com>

PingXie committed:
Update src/valkey-cli.c

Co-authored-by: Viktor Söderqvist <viktor.soderqvist@est.tech>
Signed-off-by: Ping Xie <pingxie@google.com>

PingXie commented:
I think my previous fix (https://github.com/redis/redis/pull/13055) introduced a race condition. More specifically, I don't think it is a good idea to move the slots on L3015, Instead, I should've always routed the topology update through `clusterUpdateSlotsConfigWith`. As a result, the logic attempting to fix the import source during a source primary failover is bypassed if the old primary (new replica's) `PONG` message races ahead of the new primary (old replica)'s.

![image](https://github.com/valkey-io/valkey/assets/11568491/ee869575-14be-40dd-9e6b-2d067f8e19e0)

> @PingXie I updated the PR description. Please check if it's correct and edit again if it's not.

Thanks @zuiderkwast! I will write the proper PR description next.

madolson commented on src/cluster_legacy.c:
```diff
@@ -6532,3 +6866,55 @@ int clusterAllowFailoverCmd(client *c) {
+
+/* Replicate all migrating and importing slot states to a
+ * a given replica client. */
+void clusterReplicateOpenSlots(void)
```
Will drop my comment here, since this is roughly the point we were arguing over last time. The point of contention was if we should replicate this slot ownership information by the RDB stream (the replica will know what slots are being imported based on the replication stream) vs if it should be propagated exclusively through the clustering system (In that we should be exchanging importing metadata on the clusterbus). 

@soloestoy I suppose you were the main proponent of the clusterbus approach last time. The old core team decided to go with replication based approach, but since we have new folks we should re-evaluate our options.

PingXie replied:
>I suppose you were the main proponent of the clusterbus approach last time. 

A bit more context. IIRC, @soloestoy's idea was to *replicate* the migrating/importing states via the cluster bus; while this PR *replicates* the two states as if they were the normal data.

PingXie commented on src/cluster_legacy.c:
```diff
@@ -6183,41 +6430,128 @@ int clusterCommandSpecial(client *c) {
             }
+        } else if (!strcasecmp(c->argv[3]->ptr,"node") &&
+                  c->argc == 6 &&
+                !strcasecmp(c->argv[5]->ptr,"replicaonly"))
```
I would like to revisit the addition of this "replicaonly" flag as well with the new core team. The primary rationale for going with this "replicaonly" flag at the time was because there was no precedent in a server initiated wait. I was not convinced by this argument. I think this flag is a band-aid solution and exposes the internal details unnecessarily. Public interfaces are not the right place for hacks and compromises. 

On the other hand, we could've implemented the HA contract of this slot ownership finalization logic completely in the server **without** any change on the client side and with a change that I think is quite scoped. 

How about I revise this PR with a server side implementation for comparison? 

@valkey-io/core-team

zuiderkwast replied:
I prefer server-initiated wait, i.e. blocking SETSLOT.

It's an admin command and as you say, it's better to **not** expose the internals and avoid client changes.

PingXie replied:
>It's an admin command and as you say, it's better to expose the internals and avoid client changes.

I assume you meant **not** expose? 

Yep. Let me work on a sever-initiated wait solution. I think I can come up with a cleaner solution than what I had previously.

Commit:
Consolidate cluster topology update to clusterUpdateSlotsConfigWith

Signed-off-by: Ping Xie <pingxie@google.com>

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -2985,8 +2984,9 @@ int clusterProcessPacket(clusterLink *link) {
-        /* Update the sender configEpoch if it is publishing a newer one. */
-        if (senderConfigEpoch > sender->configEpoch) {
+        /* Update the sender configEpoch if it is a primary publishing a newer one. */
+        if (!memcpy(hdr->slaveof, CLUSTER_NODE_NULL_NAME, sizeof(hdr->slaveof)) &&
```
Typo? Should be memcmp?

PingXie replied:
good catch! 👍

Commit:
Fix a typo and move gossip packet validation logic into a separate function

Signed-off-by: Ping Xie <pingxie@google.com>

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -2691,49 +2865,38 @@ static clusterNode *getNodeFromLinkAndMsg(clusterLink *link, clusterMsg *hdr) {
- * processing lead to some inconsistency error (for instance a PONG
- * received from the wrong sender ID). */
-int clusterProcessPacket(clusterLink *link) {
+bool clusterValidatePacket(clusterLink *link) {
```
Surprise!

I haven't seen bool in valkey before. Not sure why. I think we can start using it. Need to include bool.h explicitly.

The function name can be clusterIsValidPacket instead, to make the meaning of the return value more clear.

zuiderkwast commented:
Two commits need sign off. They're called "Update src/valkey-cli.c".

Commit:
Fix build builds and flaky tests

Signed-off-by: Ping Xie <pingxie@google.com>

PingXie commented:
> I think my previous fix ([redis/redis#13055](https://github.com/redis/redis/pull/13055)) introduced a race condition. More specifically, I don't think it is a good idea to move the slots on L3015, Instead, I should've always routed the topology update through `clusterUpdateSlotsConfigWith`. As a result, the logic attempting to fix the import source during a source primary failover is bypassed if the old primary (new replica's) `PONG` message races ahead of the new primary (old replica)'s.
> 
> ![image](https://private-user-images.githubusercontent.com/11568491/324680384-ee869575-14be-40dd-9e6b-2d067f8e19e0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQyNzg1ODgsIm5iZiI6MTcxNDI3ODI4OCwicGF0aCI6Ii8xMTU2ODQ5MS8zMjQ2ODAzODQtZWU4Njk1NzUtMTRiZS00MGRkLTllNmItMmQwNjdmOGUxOWUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI4VDA0MjQ0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFmODNiZGE0MTZiOWZkMzdhYzU2NjFjMzQ4MDY1ZWQ2MWZiYmUzNDYxNDJiZTYyZWZhMzFiN2U2MzI0ZTFkMDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.fb9SoVaklYyqv1xJm0K4BSPmnMRPIbRaq-LTOu2q-W8)
> 

This should now be fixed. I don't think we need to backport the change to Valkey 7.2. The original fix should still work. It is just that the new reliability improvements in this PR require all topology updates be done in `clusterUpdateSlotsConfigWith`.

Next step is to (re)introduce the server-initiate wait and get rid of the `REPLICAONLY` flag.

Commit:
Reintroduce server-initiated wait and remove the `REPLICAONLY` flag for
`CLUSTER SETSLOT NODE`

Signed-off-by: Ping Xie <pingxie@google.com>

hpatro commented on src/blocked.c:
```diff
@@ -597,23 +601,30 @@ static void handleClientsBlockedOnKey(readyList *rl) {
-/* block a client due to wait command */
-void blockForReplication(client *c, mstime_t timeout, long long offset, long numreplicas) {
+/* block a client for replica acknowledgement */
+void blockClientForReplicaAck(client *c, mstime_t timeout, long long offset, long numreplicas, int blockType, int numlocal) {
```
nit: either use `blocktype` or `block_type`.

hpatro replied:
I see we are introducing camel case for some more variables down below as well. Might be good to apply the clang formatting on thisPR.

PingXie replied:
Good catch. Will use the snake casing.

zuiderkwast replied:
If I'm not mistaken, I think the convention is camel case for prameters and snake case for local vars.

PingXie replied:
>If I'm not mistaken, I think the convention is camel case for prameters and snake case for local vars.

this doesn't seem like the case from a quick check. lots of them are neither snake nor camel, such as `tokey`, `endtime`, `dbid`, etc. I will go with `btype` as a one-off in this PR. we can have a separate discussion on the overall coding style.

![image](https://github.com/valkey-io/valkey/assets/11568491/7a547467-8bdc-4624-84c3-3cc06a827da0)

zuiderkwast replied:
You're right. Sorry!

hpatro commented on src/cluster_legacy.c:
```diff
@@ -6165,61 +6416,128 @@ int clusterCommandSpecial(client *c) {
+                blockForPreReplication(c, mstime()+1000, server.master_repl_offset+1, myself->numslaves);
+                replicationRequestAckFromSlaves();
+
+                /* Don't reply to the client yet */
```
Would be good to add comment, why we don't reply to the client in this case and when/where does it get sent to complete the command. I guess the client would be hung till then.

PingXie replied:
Yes - the client will be hung. Will do.

hpatro commented on src/cluster_legacy.c:
```diff
@@ -6532,3 +6851,55 @@ int clusterAllowFailoverCmd(client *c) {
 }
+
+/* Replicate all migrating and importing slot states to a
+ * a given replica client. */
```
Do we broadcast to all replica(s)? I don't see any replica client.

PingXie replied:
this is on the full sync path. we only need to catch the replica up on the full sync path.

zuiderkwast replied:
So we replicate to all clients that need it then? Not "a given"

```suggestion
 * replica clients. */
```

There's an "a" on the line above too....

PingXie replied:
> Do we broadcast to all replica(s)? I don't see any replica client.

Take it back - you are right that this replicates to all replicas. Will update the comment accordingly.

hpatro commented on src/cluster_legacy.c:
```diff
@@ -942,6 +942,14 @@ static void updateShardId(clusterNode *node, const char *shard_id) {
     }
 }
 
+static int areInSameShard(clusterNode *node1, clusterNode *node2) {
+    return memcmp(node1->shard_id, node2->shard_id, CLUSTER_NAMELEN) == 0;
+}
+
+static uint64_t nodeEpoch(clusterNode *n) {
+    return n->slaveof ? n->slaveof->configEpoch : n->configEpoch;
+}
```
mark these as inline?

hpatro commented on .gitignore:
```diff
@@ -41,3 +41,4 @@ Makefile.dep
 compile_commands.json
 redis.code-workspace
 .cache
+.cscope.*
```
interested to see your dev setup :D

PingXie replied:
Vim + ctags + cscope :)

madolson replied:
Just generally, we should avoid such unrelated changes ideally.

hpatro commented on src/cluster_legacy.c:
```diff
@@ -6532,3 +6851,55 @@ int clusterAllowFailoverCmd(client *c) {
+            if (nodes_ptr[j] == NULL) continue;
+
+            argv[2] = createStringObjectFromLongLongForValue(j);
+            sds name = sdsnewlen(nodes_ptr[j]->name, sizeof(nodes_ptr[j]->name));
```
This feels pretty redundant, shall we loop over each node (primary) and generate the `CLUSTER SETSLOT` command for the slot(s) owned by it.

PingXie replied:
The idea here is to replicate the "migrating/importing" states. If the slot ownership is handled via `clusterUpdateSlotsConfigWith`.

PingXie replied:
We will take a different approach in a follow up PR (#419)

hpatro commented on src/replication.c:
```diff
@@ -3637,6 +3640,7 @@ void processClientsWaitingReplicas(void) {
 
         client *c = ln->value;
         int is_wait_aof = c->bstate.btype == BLOCKED_WAITAOF;
+        int is_wait_prerepl = c->bstate.btype == BLOCKED_WAIT_PREREPL;
```
We should update the top level comment about handling the new btype state.

hpatro commented on src/cluster_legacy.c:
```diff
@@ -2453,8 +2620,16 @@ void clusterUpdateSlotsConfigWith(clusterNode *sender, uint64_t senderConfigEpoc
          * we need to remove all the keys from the slots we lost. */
-        for (j = 0; j < dirty_slots_count; j++)
+        for (int j = 0; j < dirty_slots_count; j++) {
+            serverLog(LL_WARNING,
```
Lower it to `LL_DEBUG` ?

PingXie replied:
Will lower it to NOTICE. I think the convention is to reserve `WARNING` for errors that are actionable by the "operator". Since this is a potential data loss situation, I think it warrants `NOTICE`.

hpatro reviewed (commented):
Mostly nit-picks. Pretty difficult to wrap my head around these complex conditions around slot ownership. Will re-read the code few times :D. 

* We should apply the formatting (planned to be added) on the new code changes. I see parantheses opening, spacing after `,` different from expected in few places.
* We are using primary/replica in certain place(s) and master/slave terminology in few place(s) in code comments. Maybe just use `primary/replica`.

PingXie commented:
Thanks for the review, @hpatro. There is definitely inconsistency in this PR as it has gone through too many rounds of refactoring. I will clean up the naming and comments as part of the timeout enhancement but I will leave the coding style and terminology (primary/replica) to separate PRs which I plan to address across the entire codebase.

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -6532,3 +6851,55 @@ int clusterAllowFailoverCmd(client *c) {
+    int argc = 5;
+    robj **argv = zmalloc(sizeof(robj*)*argc);
+
+    sds cmd = sdsnew("CLUSTER");
+    sds subcmd = sdsnew("SETSLOT");
+    sds importing = sdsnew("IMPORTING");
+    sds migrating = sdsnew("MIGRATING");
+    robj *cmd_obj = createObject(OBJ_STRING, cmd);
+    robj *subcmd_obj = createObject(OBJ_STRING, subcmd);
+    robj *imp_obj = createObject(OBJ_STRING, importing);
+    robj *mig_obj = createObject(OBJ_STRING, migrating);
```
The sds objects are used only once, right?

```suggestion
    robj *cmd_obj = createObject(OBJ_STRING, sdsnew("CLUSTER"));
    robj *subcmd_obj = createObject(OBJ_STRING, sdsnew("SETSLOT"));
    robj *imp_obj = createObject(OBJ_STRING, sdsnew("IMPORTING"));
    robj *mig_obj = createObject(OBJ_STRING, sdsnew("MIGRATING"));
```

If we drop the sds variables, we can drop the `_obj` suffix on the objects (if you want).

PingXie committed:
Merge branch 'valkey-io:unstable' into slot_migration

madolson committed:
Initial PR outlining the governance for the project (#345)

Initial PR to add a governance doc outlining permissions for the main
Valkey project as well as define responsibilities for sub-projects.

---------

Signed-off-by: Madelyn Olson <madelyneolson@gmail.com>
Co-authored-by: Viktor Söderqvist <viktor.soderqvist@est.tech>
Co-authored-by: Ping Xie <pingxie@outlook.com>
Co-authored-by: zhaozhao.zz <zhaozhao.zz@alibaba-inc.com>
Co-authored-by: hwware <wen.hui.ware@gmail.com>
Co-authored-by: binyan <binbin.yan@nokia.com

PingXie commented:
Quick poll on the new timeout parameter.

We talked about allowing the client to control the wait time (with a default of 5s) on a per command basis for "CLUSTER SETSLOT", which now pre-replicates the command to the replicas first. QThis new parameter can only be added to the end of the command to maintain backward compat. I think a cleaner change would be introducing a "timeout" token such as "timeout 1000" but looks like this is not a pattern used in existing commands like "WAIT" where the timeout parameter is expected to be at a fixed location. Thoughts?

hpatro commented:
For newer commands e.g. https://valkey.io/commands/xread/, we went with what you're suggesting above i.e. first the keyword and then followed by the actual value. The parsing logic becomes a slightly more complex but gives more flexibility. I think it's reasonable to proceed with your suggestion.

zuiderkwast commented:
I agree with keyword. Go for `TIMEOUT ms`.

Virusuki committed:
Modify mem_freed variable in evict.c and Update debug.c (#376)

Fix the mem_freed variable to be initialized with init.
with this PR prevents the variable from acting unknowingly.

Signed-off-by: NAM UK KIM <namuk2004@naver.com>

zhulipeng committed:
Delete unused declaration (#400)

Delete unused declaration `void *dictEntryMetadata(dictEntry *de);` in
dict.h.

---------

Signed-off-by: Lipeng Zhu <lipeng.zhu@intel.com>

zuiderkwast committed:
Don't include config.h from serverassert.h (#404)

Serverassert is a drop-in replacement of assert. We use it even in code
copied from other sources. To make these files usable outside of Valkey,
it should be enough to replace the `serverassert.h` include with
`<assert.h>`. Therefore, this file shouldn't have any dependencies to
the rest of the valkey code.

---------

Signed-off-by: Viktor Söderqvist <viktor.soderqvist@est.tech>

josiahcarlson committed:
CRC64 perf improvements from Redis patches (#350)

Improve the performance of crc64 for large batches by processing large
number of bytes in parallel and combining the results.

## Performance 
* 53-73% faster on Xeon 2670 v0 @ 2.6ghz
* 2-2.5x faster on Core i3 8130U @ 2.2 ghz
* 1.6-2.46 bytes/cycle on i3 8130U
* likely >2x faster than crcspeed on newer CPUs with more resources than
a 2012-era Xeon 2670
* crc64 combine function runs in <50 nanoseconds typical with vector +
cache optimizations (~8 *microseconds* without vector optimizations, ~80
*microseconds without cache, the combination is extra effective)
* still single-threaded
* valkey-server test crc64 --help (requires `make distclean && make
SERVER_TEST=yes`)

---------

Signed-off-by: Josiah Carlson <josiah.carlson@gmail.com>
Signed-off-by: Madelyn Olson <madelyneolson@gmail.com>
Co-authored-by: Viktor Söderqvist <viktor.soderqvist@est.tech>
Co-authored-by: Madelyn Olson <madelyneolson@gmail.com>

Commit:
Add a timeout argument to `CLUSTER SETSLOT`

Signed-off-by: Ping Xie <pingxie@google.com>

Commit:
Fix a memory corruption bug

Signed-off-by: Ping Xie <pingxie@google.com>

Commit:
Restore CMD_MAY_REPLICATE

Signed-off-by: Ping Xie <pingxie@google.com>

Shivshankar-Reddy committed:
Rename redis in aof logs and proc title redis-aof-rewrite to valkey-aof-rewrite (#393)

Renamed redis to valkey/server in aof.c serverlogs.

The AOF rewrite child process title is set to "redis-aof-rewrite" if
Valkey was started from a redis-server symlink, otherwise to
"valkey-aof-rewrite".

This is a breaking changes since logs are changed.

Part of #207.

---------

Signed-off-by: Shivshankar-Reddy <shiva.sheri.github@gmail.com>

Commit:
Reduce memory allocations

Signed-off-by: Ping Xie <pingxie@google.com>

PingXie commented:
I have incorporated all the feedback and updated the PR description. Let's merge?

zuiderkwast commented on src/commands/cluster-setslot.json:
```diff
@@ -10,7 +10,8 @@
             "ADMIN",
-            "STALE"
+            "STALE",
+            "MAY_REPLICATE"
         ],
         "arguments": [
             {
```
The optional `TIMEOUT` argument needs to be added in "arguments" below.

Is this the syntax of this command?

    CLUSTER SETSLOT slot { IMPORTING node-id | MIGRATING node-id | NODE node-id | STABLE } [ TIMEOUT timeout ]

Can you add the full syntax in the PR description?

zuiderkwast replied:
Please fix before merging.

zuiderkwast replied:
Need to add a "history" entry for this too, since it's adding a new parameter.

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -2387,21 +2409,144 @@ void clusterUpdateSlotsConfigWith(clusterNode *sender, uint64_t senderConfigEpoc
+                         * of deliberate operator actions. We should clear the importing
+                         * state to conform to the operator's will. */
+                        serverLog(LL_NOTICE,
+                            "Slot %d is not longer being imported from node %.40s (%s) in shard %.40s.",
```
```suggestion
                            "Slot %d is no longer being imported from node %.40s (%s) in shard %.40s.",
```

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -2929,8 +3121,11 @@ int clusterProcessPacket(clusterLink *link) {
+                serverLog(LL_DEBUG,
+                    "PONG contains mismatching sender ID. About node %.40s (%s) in shard %.40s added %d ms ago, having flags %d",
                     link->node->name,
+                    link->node->human_nodename,
```
What's the default value for human nodename? I'm thinking that if it hasn't been set explicitly, it may be nonsense and not good to display it in log messages like this..?

madolson replied:
It's empty by default. In a past PR we decided it's better to just show an empty () instead of having if/else everywhere. The alternative was we could also have a single node identifier that is cached in the cluster_node that is printed here.

zuiderkwast replied:
OK, if it's printed like that in other places.

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -5986,6 +6226,245 @@ char *clusterNodeGetShardId(clusterNode *node) {
+    /* Enhance cluster topology change resilience against primary failures by
+     * replicating SETSLOT before execution.
+     *
+     * Cluster topology changes such slot ownership and migrating states must
+     * be replicated to replicas before applying them to the primary. This
+     * guarantees that after a command is successfully executed, the new state
+     * won't be lost due to a primary node failure. The following example
+     * illustrates how a cluster state can be lost during slot ownership
+     * finalization:
```
Many of these comments are perfect material for a doc PR.

`topics/cluster-tutorial.md` and maybe also `topics/cluster-spec.md`. If you start a doc PR, it can be useful also for us to read in parallel when reviewing this PR.

You could even use the good old style "Starting with Valkey 8.0, bla bla bla...".

zuiderkwast replied:
https://github.com/valkey-io/valkey-doc/issues/76

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -5986,6 +6226,245 @@ char *clusterNodeGetShardId(clusterNode *node) {
+         * 2. The repl offset target is set to the master's current repl offset + 1.
+         *    There is no concern of partial replication because replicas always
+         *    ack the repl offset at the command boundary. */
+        if (timeout_ms == 0) {
+            timeout_ms = CLUSTER_OPERATION_TIMEOUT;
+        }
+        blockForPreReplication(c, mstime()+timeout_ms, server.master_repl_offset+1, myself->numslaves);
+        replicationRequestAckFromSlaves();
```
So it's not possible to bypass the wait by calling `CLUSTER SETSLOT` with `TIMEOUT 0` then? I'm fine with this behaviour. I just think it's worth noting.

PingXie replied:
skipping the wait essentially means disabling this feature. I am not sure why we would like to allow it but I am open to a discussion. is there a use case you can think of? 

> I just think it's worth noting.

Will do

zuiderkwast replied:
> is there a use case you can think of?

no

madolson replied:
If a node is disconnected might be a valid use case? If a replica is unavailable for some reason, you might still want this to through. Maybe you're trying to move the data from an unhealthy to a healthy shard by migrating the data.

zuiderkwast replied:
Yeah, anyway if you do `TIMEOUT 1` it waits 1ms but if you do `TIMEOUT 0` it waits a few seconds. Perhaps not what you'd expect.

zuiderkwast replied:
It's better that TIMEOUT 0 doesn't wait at all. It's more logical. Follow-up issue?

zuiderkwast replied:
Follow-up #422 opened to track this. Feel free to resolve and merge without it.

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -6096,130 +6575,7 @@ int clusterCommandSpecial(client *c) {
-        int slot;
-        clusterNode *n;
-
-        if (nodeIsSlave(myself)) {
-            addReplyError(c,"Please use SETSLOT only with masters.");
```
Worth noting that replicas of older versions will get here when replicating from a Valkey 8 node. IIRC, this error logged on the replica, but it may be worth checking and maybe mention in the PR description and/or docs.

madolson replied:
At some point in the future we want the replica to crash when it encounters an error from it's primary, or at least I wrote https://github.com/valkey-io/valkey/blob/8abeb79f52cbf02d16d546f48e16592173593132/valkey.conf#L757 at one point.

zuiderkwast replied:
So this will actually crash replicas with this config set to panic?

In that case, I think we should use a cluster node flag as a capability flag to indicate support for this feature. If the replica doesn't have this flag set, we don't replicate SETSLOT to it.

It'd be similar to the flag `CLUSTER_NODE_EXTENSIONS_SUPPORTED`.

zuiderkwast replied:
I opened follow-up issue #421 for this. Now we can merge without it. :)

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -6532,3 +6889,41 @@ int clusterAllowFailoverCmd(client *c) {
+    int argc = 5;
+    robj **argv = zmalloc(sizeof(robj*)*argc);
+
+    argv[0] =shared.cluster;
```
```suggestion
    argv[0] = shared.cluster;
```

zuiderkwast commented on tests/cluster/tests/21-many-slot-migration.tcl:
```diff
@@ -1,10 +1,5 @@
 # Tests for many simultaneous migrations.
 
-# TODO: Test is currently disabled until it is stabilized (fixing the test
-# itself or real issues in the server).
-
-if {false} {
```
Being able to re-enable these tests is a really good indicator that this improvement is good stuff.

zuiderkwast reviewed (commented):
High level approve. This is a very difficult PR to view in detail. There are so many corner cases. I don't know if it would have been easier if it were split up in smaller PRs, but I understand it would be too difficult to do that.

This is a true gem for release notes and we also need updated docs, so it would be good to work on those.

Have you checked if replicas would reply with -ASK redirects during migrations? If it does, that's another gem. It makes reading from replicas much more consistent than before.

hwware commented on src/server.h:
```diff
@@ -415,6 +416,7 @@ typedef enum blocking_type {
     BLOCKED_ZSET,    /* BZPOP et al. */
     BLOCKED_POSTPONE, /* Blocked by processCommand, re-try processing later. */
     BLOCKED_SHUTDOWN, /* SHUTDOWN. */
+    BLOCKED_WAIT_PREREPL, /* WAIT for pre-replication and then run the command. */
```
I found the words pre-replication is shown in several places, but I do not understand what it is (sorry missed your previous pr), Can you add a more detail explanations here so everyone can know its meaning after some time later, Thanks

PingXie replied:
will address in the follow up PR

madolson commented on src/cluster_legacy.c:
```diff
@@ -46,6 +47,7 @@
 #include <sys/stat.h>
 #include <math.h>
 #include <sys/file.h>
+#include <stdbool.h>
```
I'm going to be opinionated and say can we please not add bool now.

PingXie replied:
I like bool as it makes the function contract a lot clearer. otherwise I will have to read either the code comment or the actual function implementation.

zuiderkwast replied:
This feels almost like a separate major decision since it's so deeply rooted in this code base. We have atomics and other C99 and later features, but not bool and C++ comments. :smile:

I'd like to hear arguments against adding it though. It's part of C nowdays and it's odd to have a rule that forbids it.

PingXie replied:
sgtm. I will revert this change then. will open a separate issue to track that discussion.

madolson replied:
I'm not strictly against bool, but if we use it I would prefer we use it everywhere. I mentioned on the other thread we had an issue since someone changed the type of something from int to bool and it broke our clusterbus (it was really nasty).

zuiderkwast replied:
Discussion moved to #413.

PingXie replied:
reverted `bool` changes

madolson commented on src/cluster_legacy.c:
```diff
@@ -6532,3 +6889,41 @@ int clusterAllowFailoverCmd(client *c) {
+            argv[3] = shared.migrating;
+        }
+
+        for (int j = 0; j < CLUSTER_SLOTS; j++) {
```
I know we discussed it, but I forgot the discussion. I still think it's odd that we are putting the replica online before it has the slot information, since it will have a point in time but won't serve any data. It also is getting added to the replication buffer, which could cause replicas to unnecessarily be unable to psync. My previous PoV was why not inject this into the RDB file, which is the true point in time which we are trying to stick on at the end. We already stick "extra" metadata into the RDB file like the psync offset. If a replica comes up, it will also have a partial state.

PingXie replied:
Yeah I remember that discussion too. I had a concern back then about having two sources of the same data (the importing/migrating) with one from nodes.conf and the other from RDB. That said, I now think we could give the states in RDB higher priority. I agree closing this gap is very appealing. Will incorporate this change next.

zuiderkwast replied:
Follow-up PR?

PingXie replied:
#419

madolson commented on src/blocked.c:
```diff
@@ -597,23 +601,30 @@ static void handleClientsBlockedOnKey(readyList *rl) {
+
+/* block a client due to wait command */
+void blockForReplication(client *c, mstime_t timeout, long long offset, long numreplicas) {
+    blockClientForReplicaAck(c, timeout, offset, numreplicas, BLOCKED_WAIT, 0);
```
Do we need two separate blocked states, can we instead just use the CLIENT_PENDING_COMMAND flags?

PingXie replied:
CLIENT_PENDING_COMMAND is not a blocked state. that said, I do feel that we have a proliferation of blocked/wait states, like `wait`, `waitaof`, and now `wait_prerepl`. let me see if there is a way to improve this in the next PR. let's leave this comment open.

madolson replied:
I agree. I'm saying we already have the notion that a client is blocked with a pending command, that can get re-executed once the blocking is done. (That is how the normal command blocking works). The blocking mechanic is functionally the same as the other wait (wait for a replack from k replicas then unblock).

PingXie replied:
I agree. there is definitely room for improvement. filed #427 to track this improvement.

zuiderkwast replied:
Hehe, #427 looks so vague I doubt it will be done. :)

madolson reviewed (commented):
Didn't look through cluster_legacy yet.

madolson commented on tests/unit/cluster/slot-migration.tcl:
```diff
@@ -0,0 +1,363 @@
+        assert {$duration > 2000}
+
+        # Setslot should fail with not enough good replicas to write after the timeout
+        assert_equal {NOREPLICAS Not enough good replicas to write.} $e
```
So, in this case the slot on one of the replica will be marked as importing right? Does that ever get cleared?

PingXie replied:
>in this case the slot on one of the replica will be marked as importing right? 

correct

>Does that ever get cleared?

No. I am not implementing 2PC here. The slot migration protocol itself will never be self-sufficient. The user needs to make sure to set `importing` on the target first and only upon successful return set `migrating` on the source. This way, we can guarantee that a correct forwarding link is maintained all the time and this becomes only a performance issue (due to unnecessary forwarding) but not a correctness issue. More specifically

1. we have partial `importing` states on the target shard

 no impact because without a matching `migrating` state on the source, the forwarding won't occur
 
 2. we have partial `migrating` states on the source shard
 
this implies that a failure would be returned to the caller for its `cluster setslot` call. if a failover occurs afterwards, we would establish a forwarding link unexpectedly (to the caller). however, I think this is still fine because there would be no keys in the target shard hence a `get` request would fail in the same way (except the additional network hop). the `set` request though would be, again unexpectedly, forwarded to the target and the new key end up on the target shard. This is not desired but should be acceptable. it is mostly correct behavior with just one violation of the contract (lying about the `cluster setslot` failure`). 

also in the future when we introduce atomic slot migration, I would like the engine to automatically resume the incomplete slot migration and we will not end up in this intermediate state forever. we don't have to make a decision on the atomic slot migration behavior here but just wanted to share some additional thoughts.

madolson replied:
Yeah, it's more weird behavior, but for now it seems acceptable.

zuiderkwast replied:
We'll have to accept it until atomic slot migration, I suppose.

zuiderkwast replied:
Wait a second. Now the command fails on timeout? From earlier discussions, I though this was best effort.

What will `valkey-cli` do if SETSLOT MIGRATING or IMPORING fails?

What will `valkey-cli` do if SETSLOT NODE fails?

Is the slot migration finished or not?

madolson commented on tests/unit/cluster/slot-migration.tcl:
```diff
@@ -0,0 +1,363 @@
+        catch {R 3 DEBUG SEGFAULT} e
+
+        # Setslot with an explicit 1ms timeoout
+        set startTime [clock milliseconds]
```
```suggestion
        set start_time [clock milliseconds]
```
I think we use snake_case fairly consistently in the TCL tests?

madolson reviewed (commented):
I didn't make a lot of progress in cluster_legacy, will resume tomorrow.

simkusr committed:
Changed links and naming to valkey instead of redis (#389)

This is a minor change where only naming and links now points properly
to valkey.

Fixes #388

---------

Signed-off-by: Rolandas Šimkus <rolandas@simkus.io>
Signed-off-by: simkusr <rolandas.s@wilibox.com>
Signed-off-by: simkusr <rolandas@simkus.io>
Signed-off-by: Viktor Söderqvist <viktor.soderqvist@est.tech>
Co-authored-by: simkusr <rolandas.s@wilibox.com>
Co-authored-by: Viktor Söderqvist <viktor.soderqvist@est.tech>

zuiderkwast committed:
Document the commands JSON files (#403)

These JSON files were originally not intended to be used directly, since
they contain internals and some fiels like "acl_categories" that are not
the final ACL categories. (Valkey will apply some implicit rules to
compute the final ACL categories.) However, people see JSON files
and use them directly anyway.

So it's better to document them.

In a later PR, we can get rid of all implicit ACL categories and instead
populate them explicitly in the JSON files. Then, we'll add a validation
(e.g. in generate-command-code.py) that the implied categories are set.

---------

Signed-off-by: Viktor Söderqvist <viktor.soderqvist@est.tech>
Co-authored-by: Binbin <binloveplay1314@qq.com>

eltociear committed:
Fix typo in comment in quicklist.h (#416)

Signed-off-by: Ikko Eltociear Ashimine <eltociear@gmail.com>

Commit:
Incorporate review feedback

Signed-off-by: Ping Xie <pingxie@google.com>

zuiderkwast commented on src/commands/cluster-setslot.json:
```diff
@@ -45,6 +52,10 @@
+            },
+            {
+                "name": "timeout",
+                "type": "integer"
```
These are used for generating the syntax on the command docs.

This JSON structure is now documented (since today) here: https://github.com/valkey-io/valkey/blob/unstable/src/commands/README.md. (I realize I didn't know what "display" was but it's the lowercase *timeout* in `[ TIMEOUT timeout ]`)

```suggestion
                "display": "timeout",
                "type": "integer",
                "token": "TIMEOUT",
                "optional": true,
                "since": "8.0.0"
```

PingXie committed:
Update src/commands/cluster-setslot.json

Signed-off-by: Ping Xie <pingxie@google.com>
Co-authored-by: Viktor Söderqvist <viktor.soderqvist@est.tech>

Commit:
update commands.def

Signed-off-by: Ping Xie <pingxie@google.com>

PingXie committed:
Merge branch 'valkey-io:unstable' into slot_migration

madolson commented on src/cluster_legacy.c:
```diff
@@ -2387,21 +2408,144 @@ void clusterUpdateSlotsConfigWith(clusterNode *sender, uint64_t senderConfigEpoc
+                server.cluster->importing_slots_from[j] = NULL;
+                /* Take over the slot ownership if I am not the owner yet*/
+                if (server.cluster->slots[j] != myself) {
+                    /* We intentionally avoid updating myself's configEpoch when
+                    * taking ownership of this slot. This approach is effective
+                    * in scenarios where my primary crashed during the slot
+                    * finalization process. I became the new primary without
+                    * inheriting the slot ownership, while the source shard
+                    * continued and relinquished the slot.
```
I feel very uncomfortable about just breaking core assumptions about the algorithm here (specifically omitting the configEpoch).  It really feels like we should be bumping the epoch here, even if it means we steal slots unnecessarily. (This is an edge case after all) We are explicitly prioritizing availability here.

PingXie replied:
I think it is equally incorrect, if not more, to bump the config epoch without going through consensus. The impact on availability is the same, bumping the epoch or not, because this node will assume the ownership of this slot. In this case, the more likely case will be what I explained in the comments that follow (copied below too)

```c
                    * By not increasing myself's configEpoch, we ensure that
                    * if the slot is correctly migrated to another primary, I
                    * will not mistakenly claim ownership. Instead, any ownership
                    * conflicts will be resolved accurately based on configEpoch
                    * values. */
```

madolson replied:
>I think it is equally incorrect, if not more, to bump the config epoch without going through consensus. 

This isn't true, there is nothing wrong with bumping the epoch without consensus. Consensus is the preferred way to do it to avoid unnecessary churn, but it's not a pre-requisite. 

> The impact on availability is the same, bumping the epoch or not, because this node will assume the ownership of this slot. In this case, the more likely case will be what I explained in the comments that follow (copied below too)

I agree, that is why I'm saying we should bump the epoch. There was a change to the slot ownership, that is supposed to come with an epoch bump.

PingXie replied:
>This isn't true, there is nothing wrong with bumping the epoch without consensus. Consensus is the preferred way to do it to avoid unnecessary churn, but it's not a pre-requisite.

A primary case that I am trying to handle is where the node in question becomes a primary **without** knowing that the source shard has just given up its slot ownership to its old primary since the old primary just failed. For this node to become the new primary, it would have to go through an election and get its config epoch bumped already (with a consensus). However, it is also conceivable that the source shard might have given the slot to a **different** shard and it is in this case where I still think it is safer to **not** blindly bump the config epoch because it would undo the true indent; while in the primary failure case we don't need to bump the config epoch, because it is a side product of the failover. 

Btw, in all existing cases where we bump the config epoch without consensus, we have a clear intent to take over the slot ownership and this is the key difference IMO.

madolson replied:
> Btw, in all existing cases where we bump the config epoch without consensus, we have a clear intent to take over the slot ownership and this is the key difference IMO.

You're prescribing intent to an algorithm. The epoch get's bumped every time the slot ownership is changed, because we are in a new-epoch with new state.

madolson replied:
>  I still think it is safer to not blindly bump the config epoch because it would undo the true indent;

But you're okay with blindly serving data without true intent? I suppose that is my problem with this, either we are committing to the change with the epoch bump or we shouldn't be serving traffic.

madolson commented on src/cluster_legacy.c:
```diff
@@ -5986,6 +6225,245 @@ char *clusterNodeGetShardId(clusterNode *node) {
+int clusterParseSetSlotCommand(client *c, int *slot, clusterNode **node, int *timeout_ms) {
+    int s = -1;
+    clusterNode *n = NULL;
+    int t = 0;
```
These one letter variables are much harder to read. We already error out on failure, can we not just update the output directly?

PingXie replied:
>These one letter variables are much harder to read.

for sure.

>We already error out on failure, can we not just update the output directly?

having to dereference the pointer all the time is cumbersome. also I think it is good practice to not touch the output variables on the failure path.

madolson replied:
It's just a * right? I don't care as much as long as they have useful variable names.

madolson reviewed (commented):
Just tests left for review.

madolson commented on tests/unit/cluster/slot-migration.tcl:
```diff
@@ -0,0 +1,363 @@
+}
+
+proc wait_for_role {srv_idx role} {
+    set node_timeout [lindex [R 0 CONFIG GET cluster-node-timeout] 1]
+    # wait for a gossip cycle for states to be propagated throughout the cluster
+    after $node_timeout
```
This (surprisingly) doesn't guarantee it will propagate since it's random, why not use cluster_config_consistent?

PingXie replied:
TIL :)

madolson commented on tests/unit/cluster/slot-migration.tcl:
```diff
@@ -0,0 +1,363 @@
+proc get_open_slots {srv_idx} {
+    foreach line [split [R $srv_idx CLUSTER NODES] "\n"] {
+        set line [string trim $line]
+        if {$line eq {}} continue
+        if {[regexp {myself} $line] == 0} continue
```
I think this part of the function is just `cluster_get_myself` from cluster_util. Ditto about get_cluster_role.

madolson commented on tests/unit/cluster/slot-migration.tcl:
```diff
@@ -0,0 +1,363 @@
+}
+
+# restart a server and wait for it to come back online
+proc restart_server_and_wait {server_id} {
```
Almost universally we de-schedule processes by using the pause helpers to trigger failovers. This does two things, one is that we don't have to restart the process manually later, and second is that it's similar to when a processes disappears for a bit so tests additional modes.

Is there a specific reason we are restarting here?

PingXie replied:
>Almost universally we de-schedule processes by using the pause helpers to trigger failovers.

is this `pause_process`? it kills the process and a restart is needed too. or this is something else?

madolson replied:
It just pauses the process, I was asking if we specifically need to kill it here?

PingXie replied:
I see. my intent is to simulate true failures. If I don't kill the server, I will have to use "graceful failover" to switch roles.

zuiderkwast replied:
If the server is paused (SIGPAUSE), it doesn't respond to cluster messages and this leads to an automatic failover, in the same way as if the network were broken. I think simulates a true failure.

madolson replied:
Yeah, the segfault seems to be causing more valgrind issues as well. I think we should re-evaluate this decision.

Commit:
Improve code readability

Signed-off-by: Ping Xie <pingxie@google.com>

PingXie commented:
>Yeah, it's more weird behavior, but for now it seems acceptable.

I am not sure if I agree with the "more weird behavior" statement. There is actually a coherent story to tell/document here. The `cluster setslot` command is now pre-replicated to all replicas synchronously. If one or more replicas failed to acknowledge the command, returning "Not enough good replicas to write" seems very reasonable to me. I also don't want to introduce a very specific error just for `cluster setslot`. Do you have a different error message in mind?

Commit:
Incorporate review feedback

Signed-off-by: Ping Xie <pingxie@google.com>

PingXie commented:
> > Yeah, it's more weird behavior, but for now it seems acceptable.
> 
> I am not sure if I agree with the "more weird behavior" statement. There is actually a coherent story to tell/document here. The `cluster setslot` command is now pre-replicated to all replicas synchronously. If one or more replicas failed to acknowledge the command, returning "Not enough good replicas to write" seems very reasonable to me. I also don't want to introduce a specific error just for `cluster setslot`. Do you have a different error message in mind?

Okay now I think this is "weird behavior" on the github side :). It apparently misplaced your comment and made it look like it was about the error message but after I clicked into the code, I saw that you were commenting about the "torn" states. Please ignore my response above.

zuiderkwast commented on tests/unit/cluster/slot-migration.tcl:
```diff
@@ -0,0 +1,355 @@
+        wait_for_slot_state 7 "\[609-<-$R0_id\]"
+    }
+
+    test "Empty-shard migration target is auto-updated after faiover in target shard" {
```
How will `valkey-cli` behave if it's in the middle of transferring keys between the nodes when the failover happens? WIll it seamlessly continue and finish the slot after failover?

zuiderkwast replied:
I'm starting to think that we could let a cron job could finish the (non-atomic) slot migration, i.e. transfer keys. Without `valkey-cli` involved, it might be easier to get a self-healing cluster.

zuiderkwast replied:
If we introduce a command like `CLUSTER MIGRATE slot TO node`, initially with a simple non-atomic cron-driven implementation, then we can replace it with an atomic slot migration later. It won't be a breaking change (I think).

PingXie replied:
All good points and if you remember our discussion back in the Redis days these were all on my plan

1. Introduce "cluster migrate slot"
2. Have engine drive the migration in serverCron
3. Auto resume on failures

I also believe the same interface can be reused for atomic slot migration without causing breaks.

Let's file separate issues?

zuiderkwast replied:
Yeah, incremental perfection. .)

For now, I just would like to know how this PR affects the documented `valkey-cli`-based flow. Can it resume a slot-migration? It quite possibly can. It'd just be good to test it.

PingXie replied:
>For now, I just would like to know how this PR affects the documented valkey-cli-based flow. Can it resume a slot-migration? It quite possibly can. It'd just be good to test it.

From a quick look at `clusterManagerMigrateKeysInSlot`, it does not look like that valkey-cli would be able to resume on failovers. `MIGRATE` is a write command that can only be executed on the primary but I haven't seen valkey-cli rediscovering the topology in `clusterManagerMigrateKeysInSlot`?

zuiderkwast replied:
No, `--cluster fix` maybe. If not, I think we should probably add something there to discover and resume an aborted migration.

Commit:
Reintroduce a proper timeout to help stablize the cluster

Signed-off-by: Ping Xie <pingxie@google.com>

Commit:
update code comment

Signed-off-by: Ping Xie <pingxie@google.com>

Commit:
Bump the config epoch when taking over "orphaned" slots

Signed-off-by: Ping Xie <pingxie@google.com>

PingXie committed:
Merge branch 'valkey-io:unstable' into slot_migration

PingXie commented:
Sorry I can't fix the DCO. Shouldn't have rebased. Closing this PR and moving changes to #445.
